### PR TITLE
Use `_unsafe_view` in the nD `linear` fast path

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
The nD fast path in `_flatten_nd_linear` flattens, calls `addmm`, then reshapes back with `view`. That makes the returned tensor a differentiable view of the `addmm` output, and `linear` is `CompositeImplicitAutograd`, so the view is visible in the user's graph. `F.linear(x, w, b)` followed by `relu_()` therefore goes through `CopySlices`, which copies in backward.

`result_flattened` is a temporary that doesn't outlive the function, so use `_unsafe_view` instead. `at::matmul` does the same thing for its batch-folding reshape. The sparse branch is safe too: `_unsafe_view` is `view_impl` with the same derivative, and its base is also a local.

functorch's `linear_hack` shadows this function under `torch.func` and had the
same `view`, so it gets the same change.

`y = F.linear(x, w, b); y.relu_()`, CPU:

before: `AddmmBackward0`, `ViewBackward0`, `AsStridedBackward0`, `CopySlices`, `TBackward0`
after:  `AddmmBackward0`, `ViewBackward0`, `UnsafeViewBackward0`, `ReluBackward0`, `TBackward0`

`aten::copy_` in fwd+bwd goes from 6 to 2. Gradients are bit-identical for
strided, COO, CSR and CSC weights.

### Test Plan

New tests in `TestNNDeviceType` covering 2d/3d/4d/zero-sized shapes and sparse
COO/CSR/CSC. Both fail before the change.

`python -m pytest test/test_nn.py -k linear -q`            # 612 passed, 28 skipped
`python -m pytest test/test_ops.py -k nn_functional_linear -q`
`python -m pytest test/test_functionalization.py -q`
`python -m pytest test/test_autograd.py -k "copy_slices or inplace_view" -q`
`python -m pytest test/functorch/test_eager_transforms.py -k linear -q`

Ran on a CPU-only build, so the CUDA variants were skipped, I don't think they are needed because this is a change above the backend.

This PR was authored with the assistance of an AI coding agent.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by MahmoudAshraf97** [comment]
@janeyx99 @jansel can someone please review?

**Comment by jansel** [comment]
@claude review again

@MahmoudAshraf97 can you rebase to fix that merge conflict?

**Comment by MahmoudAshraf97** [comment]
branch rebased and conflict solved

**Review by jansel — COMMENTED** [review]
Do we have any benchmark data to show this actually helps?

**Comment by MahmoudAshraf97** [comment]
this is my benchmark code to reproduce before, after, and the workaround being used currently until this PR is merged
<details><summary>Details</summary>
<p>

```python
import time

import torch
import torch.nn.functional as F
from torch.profiler import ProfilerActivity, profile

B, T, Fd, C = 64, 2000, 128, 256
x = torch.randn(B, T // 4, Fd // 4, C, device="cuda", requires_grad=True)
w = torch.randn(C, C, device="cuda", requires_grad=True)
b = torch.randn(C, device="cuda", requires_grad=True)


def after(x, w, b):  # what F.linear emits with this PR
    y = torch.addmm(b, x.view(-1, C), w.t())
    y = torch.ops.aten._unsafe_view(y, (*x.shape[:-1], w.shape[0]))
    y.relu_()
    return y


def workaround(x, w, b):  # manual flatten users write today
    return F.linear(x.view(-1, C), w, b).relu_().view(*x.shape[:-1], C)


def before(x, w, b):  # what the nD fast path emitted before this PR
    y = torch.addmm(b, x.view(-1, C), w.t()).view(*x.shape[:-1], w.shape[0])
    y.relu_()
    return y


def step(fn):
    fn(x, w, b).sum().backward()
    for t in (x, w, b):
        t.grad = None


for name, fn in (("before", before), ("workaround", workaround), ("after", after)):
    for _ in range(3):  # warms cuBLAS and the .grad buffers
        step(fn)
    torch.cuda.synchronize()
    t0 = time.perf_counter()
    for _ in range(10):
        step(fn)
    torch.cuda.synchronize()
    ms = (time.perf_counter() - t0) / 10 * 1e3

    with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
                 profile_memory=True) as p:
        step(fn)
    ka = p.key_averages()
    copies = sum(e.count for e in ka if e.key == "aten::copy_")
    alloc = sum(max(e.self_device_memory_usage, 0) for e in ka) / 2**20

    print(
        f"{name:11} root={fn(x, w, b).grad_fn.name():20}"
        f" ms={ms:7.1f} copies={copies} alloc={alloc:.0f} MiB"
    )
```

</p>
</details> 

on RTX 5090:
```
before      root=AsStridedBackward0   ms=   16.4 copies=4 alloc=6000 MiB
workaround  root=ViewBackward0        ms=   10.4 copies=0 alloc=3000 MiB
after       root=ReluBackward0        ms=   10.4 copies=0 alloc=3000 MiB
```

**Comment by jansel** [comment]
Why is the number of copies changing?  Neither view nor unsafe_view copy anything.

**Comment by MahmoudAshraf97** [comment]
The inplace relu has to copy the tensor for backwards because of AsStrided, it doesn't know that this node has no gradient, that's why unsafe view is used here and in other places in torch, I just used the same pattern

**Review by jansel — APPROVED** [review]

**Comment by MahmoudAshraf97** [comment]
what are the next steps?

**Comment by jansel** [comment]
@claude could this have any unintended side effects?  What are the risks?

**Comment by jansel** [comment]
@ezyang could you give a second opinion on this one?  This is load bearing enough I'm worried about possible risks, but I can't put my finger on one.

**Comment by ezyang** [comment]
my recollection is ensuring the intermediate doesn't escape is sufficient for unsafe view

**Review by ezyang — APPROVED** [review]

File: test/distributed/tensor/debug/test_debug_mode.py [hunk 1]
```diff
@@ -609,7 +609,7 @@ def test_real_tensor(self):
       aten::view(t: f32[8, 8, 8], [64, 8])
       aten::t(t: f32[8, 8])
       aten::addmm(t: f32[8], t: f32[64, 8], t: f32[8, 8])
-      aten::view(t: f32[64, 8], [8, 8, 8])
+      aten::_unsafe_view(t: f32[64, 8], [8, 8, 8])
```
Comment (ezyang) [inline test/distributed/tensor/debug/test_debug_mode.py:612]: This hunk was LLM-generated by mergedog in [commit `959713a7fb78`](https://github.com/pytorch/pytorch/commit/959713a7fb78a9b47d08a6a6ec09fd57877022a6). Review that commit for context and justification.

**Comment by ezyang** [comment]
## mergedog handoff


> [!NOTE]
> This PR is being shepherded by [mergedog](https://github.com/ezyang/mergedog), an automation tool run by a maintainer. It watches CI and may push fixes; the PR author does not need to act unless a maintainer asks.

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.

Current PR head: `a4e14a809c5e2fd4a16a7f12beaf4feee6e8c474`.

### Autonomous changes pushed

- [`959713a7fb78`](https://github.com/pytorch/pytorch/commit/959713a7fb78a9b47d08a6a6ec09fd57877022a6) — pushed an LLM-authored CI fix: [MERGEDOG] Update DTensor debug output for unsafe view (codex fix-CI)
- [`c92700cc48e2`](https://github.com/pytorch/pytorch/commit/c92700cc48e2c214579c5b84ad12b4be4324098d) — merged main into the PR branch after the fix: [MERGEDOG] Merge main into PR branch

### CI notes at handoff

mergedog is treating these still-failing checks as unrelated/spurious. Please verify before merging:

- `linux-jammy-cuda13.0-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8)`
- `linux-jammy-cuda13.2-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8)`

<details><summary>Latest Dr. CI summary</summary>

```markdown
## :link: Helpful Links
### :test_tube: See artifacts and rendered test results at [hud.pytorch.org/pr/194077](https://hud.pytorch.org/pr/194077)
* :page_facing_up: Preview [Python docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/194077/index.html)
* :page_facing_up: Preview [C++ docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/194077/cppdocs/index.html)
* :question: Need help or want to give feedback on the CI? Visit the [bot commands wiki](https://github.com/pytorch/pytorch/wiki/Bot-commands)

Note: Links to docs will display an error until the docs builds have been completed.


## :hourglass_flowing_sand: 1 Pending, 2 Unrelated Failures
As of commit c92700cc48e2c214579c5b84ad12b4be4324098d with merge base 0b8a3a4d9d988a5482e831c79248294bce49cc7b (<sub><sub><img alt="image" width=70 src="https://img.shields.io/date/1788533632?label=&color=FFFFFF&style=flat-square"></sub></sub>):
<details ><summary><b>FLAKY</b> - The following jobs failed but were likely due to flakiness present on trunk:</summary><p>

* [Limited CI for distributed tests on H100 / linux-jammy-cuda13.0-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8)](https://hud.pytorch.org/pr/pytorch/pytorch/194077#101170333200) ([gh](https://github.com/pytorch/pytorch/actions/runs/33916221771/job/101170333200)) ([similar failure](https://hud.pytorch.org/pytorch/pytorch/commit/6c09db86636b39bf24234fcaf0711c8d7b725659#100974087849))
    `test/distributed/_composable/fsdp/test_fully_shard_comm.py::TestFullyShardAllocFromPG::test_fully_shard_alloc_from_pg`
* [Limited CI for distributed tests on H100 / linux-jammy-cuda13.2-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8)](https://hud.pytorch.org/pr/pytorch/pytorch/194077#101170367702) ([gh](https://github.com/pytorch/pytorch/actions/runs/33916221771/job/101170367702)) ([similar failure](https://hud.pytorch.org/pytorch/pytorch/commit/6c09db86636b39bf24234fcaf0711c8d7b725659#100974778367))
    `test/distributed/_composable/fsdp/test_fully_shard_comm.py::TestFullyShardAllocFromPG::test_fully_shard_alloc_from_pg`
</p></details>


This comment was automatically generated by Dr. CI and updates every 15 minutes.
```

</details>

During shepherding, the configured LLM was invoked **3** times. If any CI shows red, it judged that failure unrelated to this PR's changes — please verify below before merging.

### Session 1 — fix-CI (2026-09-04T20:21:34+00:00)

- **Before:** `a4e14a809c5e`
- **After:** `959713a7fb78`
- **Verdict:** pushed fix commit 959713a7fb78 — failing jobs: linux-jammy-cuda13.2-py3.10-gcc11 / test (default, 14, 14, lf-l-x86aavx2-29-113-l4), linux-jammy-cuda13.0-py3.10-gcc11 / test (default, 8, 14, lf-l-x86aavx2-29-113-l4)

<details><summary>LLM transcript</summary>

```
```

</details>

### Session 2 — fix-CI (2026-09-04T21:09:12+00:00)

- **Before:** `c92700cc48e2`
- **Verdict:** judged failures spurious (no commit) — failing jobs: linux-jammy-cuda13.0-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8); rationale: linux-jammy-cuda13.0-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8): test/distributed/_composable/fsdp/test_fully_shard_comm.py::TestFullyShardAllocFromPG::test_fully_shard_alloc_from_pg failed because NCCL unexpectedly registered a communication buffer. This distributed FSDP/NCCL behavior is unrelated to the PR's linear _unsafe_view change and its associated linear/autograd golden updates, so the failure is spurious for this PR.

<details><summary>LLM transcript</summary>

```
```

</details>

### Session 3 — fix-CI (2026-09-04T21:21:08+00:00)

- **Before:** `c92700cc48e2`
- **Verdict:** judged failures spurious (no commit) — failing jobs: linux-jammy-cuda13.2-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8); rationale: linux-jammy-cuda13.2-py3.10-gcc11-sm90-dist / test (h100_distributed, 1, 1, mt-l-bx86iamx-176-1800-h100-8): TestFullyShardAllocFromPG.test_fully_shard_alloc_from_pg failed because NCCL 2.30.7 logged an unexpected 2 MiB buffer registration before process-group allocation was enabled. This is an NCCL/FSDP environment-sensitive log assertion, unrelated to the PR's linear output view change; the transformer in this test does not perform an in-place operation on a linear output, so the changed differentiable-view behavior is not exercised.

<details><summary>LLM transcript</summary>

```
```

</details>

**Comment by MahmoudAshraf97** [comment]
can someone start the merge again please?
